Sources/tart/Commands/Push.swift

```diff
@@ -35,12 +39,21 @@ struct Push: AsyncParsableCommand {
       for (registryIdentifier, remoteNamesForRegistry) in registryGroups {
         let registry = try Registry(host: registryIdentifier.host, namespace: registryIdentifier.namespace)
 
-        let listOfTagsAndDigests = "{" + remoteNamesForRegistry.map{$0.fullyQualifiedReference }
-                .joined(separator: ",") + "}"
         defaultLogger.appendNewLine("pushing \(localName) to "
-          + "\(registryIdentifier.host)/\(registryIdentifier.namespace)\(listOfTagsAndDigests)...")
+          + "\(registryIdentifier.host)/\(registryIdentifier.namespace)\(remoteNamesForRegistry.referenceNames())...")
+
+        let manifestDigest = try await localVMDir.pushToRegistry(registry: registry, references: remoteNamesForRegistry.map{ $0.reference.value })
 
-        try await localVMDir.pushToRegistry(registry: registry, references: remoteNamesForRegistry.map{ $0.reference })
+        // Populate the local cache (if requested)
+        if populateCache {
+          for remoteName in remoteNamesForRegistry {
+            defaultLogger.appendNewLine("caching \(localName) as \(remoteName)...")
+
+            if let cacheToVMDir = try VMStorageOCI().cache(name: remoteName, digest: manifestDigest) {
+              try localVMDir.clone(to: cacheToVMDir, generateMAC: false)
+            }
+          }
+        }
```

Comment:
>And I still think there is too much cloning is going on.

It only clones if the "digest" directory doesn't exist, otherwise creates a symbolic link. This is achieved by re-using the `pull()` code.

> What if `pushToRegistry` will return a `RemoteName with the `shaXXX` then the cache population logic might be much simpler:

I see two downsides to the example you've showed:

1. It seems that the `localVMDir.clone()` will always clone.
2. Not reusing the `pull()`'s logic results in needless duplication.

Reply:
I don't think there will much "needless duplication". 🤷‍♂️ Here is my take #107